src/szk.rs

```diff
@@ -0,0 +1,469 @@
+use crate::vdaf::xof::Seed;
+#[cfg(test)]
+use crate::vdaf::xof::{IntoFieldVec, Xof, XofTurboShake128};
+#[cfg(test)]
+use std::borrow::Cow;
+#[cfg(test)]
+use std::marker::PhantomData;
+///Domain separation tags
+#[cfg(test)]
+const DST_PROVE_RANDOMNESS: u16 = 0;
+#[cfg(test)]
+const DST_PROOF_SHARE: u16 = 1;
+#[cfg(test)]
+const DST_QUERY_RANDOMNESS: u16 = 2;
+#[cfg(test)]
+const DST_JOINT_RAND_SEED: u16 = 3;
+#[cfg(test)]
+const DST_JOINT_RAND_PART: u16 = 4;
+#[cfg(test)]
+const DST_JOINT_RANDOMNESS: u16 = 5;
+// I don't know how to access the crate version number .
+#[cfg(test)]
+const VERSION: u8 = 8;
+
+/// Errors propagated by methods in this module.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+#[cfg(test)]
+pub enum SzkError {
+    /// Calling [`Type::prove`] returned an error.
+    #[error("prove error: {0}")]
+    Prove(String),
+
+    /// Calling [`Type::query`] returned an error.
+    #[error("query error: {0}")]
+    Query(String),
+
+    /// Calling [`Type::decide`] returned an error.
+    #[error("decide error: {0}")]
+    Decide(String),
+
+    /// Generic invalid parameter. This may be returned when a type cannot be constructed.
+    #[error("invalid paramter: {0}")]
+    InvalidParameter(String),
+
+    /// Returned if an FLP operation encountered an error.
+    #[error("Flp error: {0}")]
+    Flp(#[from] FlpError),
+
+    /// PRNG error.
+    #[error("prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// Codec error.
+    #[error("codec error: {0}")]
+    Codec(#[from] CodecError),
+}
+
+/// Contains an FLP proof share, and if joint randomness is needed, the blind
+/// used to derive it and the other party's joint randomness part.
+pub enum SzkProofShare<'a, const SEED_SIZE: usize, F: FftFriendlyFieldElement> {
+    ///Leader's proof share is uncompressed.
+    ///The first Seed is a blind, second is a joint randomness part.
+    Leader(Vec<F>, Option<&'a Seed<SEED_SIZE>>, Option<Seed<SEED_SIZE>>),
+    ///Helper's seed is compressed using the same blind as its joint randomness
+    Helper(&'a Seed<SEED_SIZE>, Option<Seed<SEED_SIZE>>),
+}
+
+/// A tuple of two proof shares: one for each aggregator
+pub type SzkProofShares<'a, const SEED_SIZE: usize, F> = (
+    SzkProofShare<'a, SEED_SIZE, F>,
+    SzkProofShare<'a, SEED_SIZE, F>,
+);
+/// A tuple containing the state and messages produced by an SZK query.
+pub type SzkQueryShares<const SEED_SIZE: usize, F> = (
+    Option<Seed<SEED_SIZE>>,
+    SzkVerifier<F>,
+    Option<Seed<SEED_SIZE>>,
+);
+/// Verifier type for Szk Proof
+pub type SzkVerifier<F> = Vec<F>;
+
+///Main struct encapsulating the shared zero-knowledge functionality.
+/// The type T is the underlying FLP proof system.
+/// P is the XOF used to derive all random coins (it should be indifferentiable
+/// from a random oracle for security.)
+#[cfg(test)]
+pub struct Szk<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    typ: T,
+    algorithm_id: u32,
```

Comment:
Where are these algorithm IDs assigned? Perhaps an enum would be a good idea?

Reply:
They're passed in from whatever higher-level VDAF is using the SZK (i.e., this will eventually be the Mastic algorithm ID).